### PR TITLE
connect_and_check: Allow choosing which connections to refresh.

### PR DESCRIPTION
**Pull request by nihohit**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: redis/src/cluster_async/connections_logic.rs [hunk 1]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+            .await
+            .map(|(conn, _ip)| conn)
+            .ok();
+        (user_connection, management_connection)
+    } else {
+        (prev_node.user_connection, Some(new_conn))
+    };
+
+    if let Some(new_conn) = management_connection.as_mut() {
+        // The new IP matches the existing one. Use this connection for the management connection.
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:267]: This comment should be moved to the else in line 262

File: redis/src/cluster_async/connections_logic.rs [hunk 2]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+        // TODO - should we fail here even if a new user connection was set up? or in that situation, should we return the new connection?
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:268]: Why should we fail? if we have a working user connection, I think ManagementConnectionFailed is exactly what needed. Maybe we can increase the retries for setting up the management/user connection, as it is done after a connection is already esatblished

File: redis/src/cluster_async/connections_logic.rs [hunk 3]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+                node: ClusterNode {
+                    user_connection,
+                    ip: new_ip,
+                    management_connection: None,
+                },
+                err,
+            };
+        };
+    };
+    ConnectAndCheckResult::Success(ClusterNode {
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:280]: why not using create_async_node

File: redis/src/cluster_async/connections_logic.rs [hunk 4]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+    } else {
+        (prev_node.user_connection, Some(new_conn))
+    };
+
+    if let Some(new_conn) = management_connection.as_mut() {
+        // The new IP matches the existing one. Use this connection for the management connection.
+        // TODO - should we fail here even if a new user connection was set up? or in that situation, should we return the new connection?
+        if let Err(err) = setup_management_connection(new_conn).await {
+            return ConnectAndCheckResult::ManagementConnectionFailed {
+                node: ClusterNode {
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:271]: why not using create_async_node

File: redis/src/cluster_async/connections_logic.rs [hunk 5]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+                        node: create_async_node(user_conn, None, user_ip),
+                        err: (ErrorKind::IoError, "mismatched IP").into(),
+                    },
+                    Err(err) => err.into(),
+                }
+            }
+        }
+        (Ok(conn), Err(err)) | (Err(err), Ok(conn)) => {
+            // Only a single connection was successfully established. Use it for the user connection
+            warn_management_conn_faild(addr, &err);
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:188]: seems like we're not adding this warning each time the management connection fails. maybe we can have a function that creates ManagementConnectionFailed and in the function add the warning, so each time ManagementConnectionFailed is returned we will add the warning

File: redis/src/cluster_async/connections_logic.rs [hunk 6]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+                    "Node: {:?} received errors: `{:?}`, `{:?}`",
+                    addr, err_1, err_2
+                ),
+            ))
+            .into()
+        }
+    }
+}
+
+async fn create_management_connection<C>(
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:213]: I think this function is redundant, we can add the warning into the ManagementConnectionFailed result constructor

File: redis/src/cluster_async/connections_logic.rs [hunk 7]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+                None => connect_and_check_all_connections(addr, params, socket_addr).await,
+            }
+        }
+        RefreshConnectionType::AllConnections => {
+            connect_and_check_all_connections(addr, params, socket_addr).await
+        }
+    }
+}
+
+async fn create_user_connection<C>(
```
Comment (barshaul) [inline redis/src/cluster_async/connections_logic.rs:388]: create_and_setup_management_connection and create_user_connection names should be the same in all but the connection name (user/management)

File: redis/tests/test_async_cluster_connections_logic.rs [hunk 8]
```diff
@@ -15,8 +15,34 @@ use support::{
...
     use super::*;
-    use redis::cluster_async::testing::connect_and_check;
+    use redis::cluster_async::testing::{connect_and_check, ConnectAndCheckResult};
+
+    fn assert_partial_result(
+        result: ConnectAndCheckResult<MockConnection>,
+    ) -> (AsyncClusterNode<MockConnection>, redis::RedisError) {
+        match result {
+            ConnectAndCheckResult::ManagementConnectionFailed { node, err } => (node, err),
+            ConnectAndCheckResult::Success(_) => panic!("full success"),
```
Comment (barshaul) [inline redis/tests/test_async_cluster_connections_logic.rs:30]: please make the panic messages more informative

File: redis/tests/test_async_cluster_connections_logic.rs [hunk 9]
```diff
@@ -32,19 +58,369 @@ mod test_connect_and_check {
...
+            &format!("{name}:6379"),
+            params.clone(),
+            None,
+            RefreshConnectionType::AllConnections,
+            None,
+        )
+        .await;
+        let (node, _) = assert_partial_result(result);
+        assert!(node.management_connection.is_none());
+        assert_eq!(node.ip, None);
```
Comment (barshaul) [inline redis/tests/test_async_cluster_connections_logic.rs:110]: why? seems irrelevant for this test

File: redis/tests/test_async_cluster_connections_logic.rs [hunk 10]
```diff
@@ -32,19 +58,369 @@ mod test_connect_and_check {
...
+            &format!("{name}:6379"),
+            params,
+            None,
+            RefreshConnectionType::AllConnections,
+            None,
+        )
+        .await;
+        let (node, _) = assert_partial_result(result);
+        assert!(node.management_connection.is_none());
+        assert_eq!(node.ip, None);
```
Comment (barshaul) [inline redis/tests/test_async_cluster_connections_logic.rs:128]: same

**Review by barshaul — CHANGES_REQUESTED** [review]

File: redis/src/cluster_async/connections_logic.rs [hunk 11]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
+                .await
+                .get_node();
             }
         };
         match check_connection(&mut conn, params.connection_timeout.into()).await {
-            Ok(_) => Ok((conn, node.ip)),
-            Err(_) => connect_and_check(addr, params.clone(), None).await,
+            Ok(_) => Ok(AsyncClusterNode::new(
+                async { conn }.boxed().shared(),
+                None,
```
[the remark this answers is not in the document]
Comment (shachlanAmazon) [reply redis/src/cluster_async/connections_logic.rs:79]: yup

[the remark this answers is not in the document]
Comment (shachlanAmazon) [reply redis/src/cluster_async/connections_logic.rs:79]: This PR is large enough, as it is. I don't want to make it larger than it needs to be.

File: redis/src/cluster_async/connections_logic.rs [hunk 12]
```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
...
     if let Some(node) = node {
         let mut conn = node.user_connection.await;
         if let Some(ref ip) = node.ip {
             if has_dns_changed(addr, ip).await {
-                return connect_and_check(addr, params.clone(), None).await;
+                return connect_and_check(
+                    addr,
+                    params.clone(),
+                    None,
+                    RefreshConnectionType::OnlyUserConnection,
```
[the remark this answers is not in the document]
Comment (shachlanAmazon) [reply redis/src/cluster_async/connections_logic.rs:69]: This PR tries to minimize changes to `get_or_create_conn`, and ATM we don't use management connections.

File: redis/src/cluster_async/connections_logic.rs (hunk 3 above)
Comment (shachlanAmazon) [reply redis/src/cluster_async/connections_logic.rs:280]: because `user_connection` isn't C, it's ConnectionFuture<C>

File: redis/src/cluster_async/connections_logic.rs (hunk 4 above)
Comment (shachlanAmazon) [reply redis/src/cluster_async/connections_logic.rs:271]: because user_connection isn't C, it's ConnectionFuture<C>

**Comment by shachlanAmazon** [comment]
@barshaul round

**Review by barshaul — APPROVED** [review]
create_and_setup_management_connection  -> create_management_connection (or create_user_connection -> create_and_setup_user_connection)
other than that 👌
